File: src/net/KNet/Specific/Connect/KNetSinkTask.cs
/*
*  Copyright (c) 2021-2026 MASES s.r.l.
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*  http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
*
*  Refer to LICENSE for more information.
*/

using Java.Util;
using MASES.JCOBridge.C2JBridge;
using Org.Apache.Kafka.Connect.Sink;
using System.Collections.Generic;

namespace MASES.KNet.Connect
{
    #region IKNetSinkTask
    /// <summary>
    /// Helper interface for <see cref="KNetSinkTask{TTask}"/>
    /// </summary>
    public interface IKNetSinkTask : IKNetTask
    {

    }

    #endregion

    #region KNetSinkTask<TTask>

    /// <summary>
    /// An implementation of <see cref="KNetTask{TTask}"/> for sink task
    /// </summary>
    /// <typeparam name="TTask">The class which extends <see cref="KNetSinkTask{TTask}"/></typeparam>
    public abstract class KNetSinkTask<TTask> : KNetTask<TTask>, IKNetSinkTask
        where TTask : KNetSinkTask<TTask>
    {
        /// <summary>
        /// The <see cref="Put(IEnumerable{SinkRecord})"/> uses the <see cref="JCOBridgeExtensions.WithPrefetch{TEnumerable}(TEnumerable, bool)"/>
        /// </summary>
        protected virtual bool UsePrefetch { get; set; } = false;
        /// <summary>
        /// The <see cref="Put(IEnumerable{SinkRecord})"/> uses the <see cref="JCOBridgeExtensions.WithThread{TEnumerable}(TEnumerable, bool, System.Threading.ThreadPriority)"/>
        /// </summary>
        protected virtual bool UseThread { get; set; } = false;
        /// <summary>
        /// The <see cref="System.Threading.ThreadPriority"/> to be used when <see cref="UseThread"/> is <see langword="true"/>
        /// </summary>
        protected virtual System.Threading.ThreadPriority ThreadPriority { get; set; } = System.Threading.ThreadPriority.AboveNormal;
        /// <summary>
        /// The <see cref="SinkTaskContext"/>
        /// </summary>
        public SinkTaskContext Context => Context<SinkTaskContext>();
        /// <summary>
        /// Set the <see cref="ReflectedTaskClassName"/> of the connector to a fixed value
        /// </summary>
        protected override string ReflectedTaskClassName => "KNetSinkTask";
        /// <summary>
        /// Public method used from Java to trigger <see cref="Put(IEnumerable{SinkRecord})"/>
        /// </summary>
        [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
        public void PutInternal()
        {
            try
            {
                using Collection<SinkRecord> collection = DataToExchange<Collection<SinkRecord>>();
                using var collection1 = collection.WithPrefetch(UsePrefetch).WithThread(UseThread, ThreadPriority);
                Put(collection1);
            }
            catch (System.Exception e)
            {
                LogError($"PutInternal failed with {e}");
                throw;
            }
        }
        /// <summary>
        /// Implement the method to execute the Put action
        /// </summary>
        /// <param name="collection">The set of <see cref="SinkRecord"/> from Apache Kafka Connect framework</param>
        public abstract void Put(IEnumerable<SinkRecord> collection);
    }

    #endregion
}

File: src/jvm/knet/src/main/java/org/mases/knet/developed/connect/sink/KNetSinkTask.java
/*
 *  Copyright 2024 MASES s.r.l.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Refer to LICENSE for more information.
 */

package org.mases.knet.developed.connect.sink;

import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.sink.SinkRecord;
import org.apache.kafka.connect.sink.SinkTask;
import org.apache.kafka.connect.sink.SinkTaskContext;
import org.mases.jcobridge.*;
import org.mases.knet.developed.connect.KNetConnectLogging;
import org.mases.knet.developed.connect.KNetConnectProxy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Collection;
import java.util.Map;

public class KNetSinkTask extends SinkTask implements KNetConnectLogging {
    private static final Logger log = LoggerFactory.getLogger(KNetSinkTask.class);

    private static final String registrationName = "KNetSinkTask";

    long taskId = 0;

    String indexedRegistrationName;

    JCObject sinkTask = null;

    Object dataToExchange = null;

    public Object getDataToExchange() {
        return dataToExchange;
    }

    public void setDataToExchange(Object dte) {
        dataToExchange = dte;
    }

    public KNetSinkTask() throws ConnectException, JCException, IOException {
        super();
        log.debug("Invoking ctor of KNetSinkTask");
        taskId = KNetConnectProxy.getNewTaskId();
        indexedRegistrationName = String.format("%s_%d", registrationName, taskId);
        if (JCOBridge.isCLRHostingProcess()) {
            JCOBridge.RegisterJVMGlobal(indexedRegistrationName, this);
            JCObject sink = KNetConnectProxy.getSinkConnector();
            if (sink == null) throw new ConnectException("getSinkConnector returned null.");
            sinkTask = (JCObject) sink.Invoke("AllocateTask", taskId);
        }
    }

    public SinkTaskContext getContext() {
        return context;
    }

    @Override
    public String version() {
        log.debug("Invoking version");
        try {
            if (sinkTask != null) {
                return (String) sinkTask.Invoke("VersionInternal");
            }
        } catch (JCNativeException jcne) {
            log.error("Failed Invoke of \"version\"", jcne);
        }
        return "NOT AVAILABLE";
    }

    @Override
    public void start(Map<String, String> map) {
        log.debug("Invoking start");
        try {
            if (!JCOBridge.isCLRHostingProcess()) {
                JCOBridge.RegisterJVMGlobal(indexedRegistrationName, this);
                String connectorId = KNetConnectProxy.getConnectorId(map);
                JCObject sink = KNetConnectProxy.getConnector(connectorId);
                if (sink == null) throw new ConnectException("getConnector returned null.");
                sinkTask = (JCObject) sink.Invoke("AllocateTask", taskId);
            }

            try {
                dataToExchange = map;
                sinkTask.Invoke("StartInternal");
            } finally {
                dataToExchange = null;
            }
        } catch (JCException | IOException jcne) {
            log.error("Failed Invoke of \"start\"", jcne);
        }
    }

    @Override
    public void put(Collection<SinkRecord> collection) {
        log.debug("Invoking put");
        try {
            try {
                dataToExchange = collection;
                sinkTask.Invoke("PutInternal");
            } finally {
                dataToExchange = null;
            }
        } catch (JCNativeException jcne) {
            log.error("Failed Invoke of \"put\"", jcne);
        }
    }

    @Override
    public void stop() {
        log.debug("Invoking stop");
        try {
            sinkTask.Invoke("StopInternal");
        } catch (JCNativeException jcne) {
            log.error("Failed Invoke of \"stop\"", jcne);
        }
        finally {
            if (!JCOBridge.isCLRHostingProcess()) {
                try {
                    JCOBridge.UnregisterJVMGlobal(indexedRegistrationName);
                } catch (JCNativeException e) {
                    throw new RuntimeException(e);
                }
            }
        }
    }

    @Override
    public boolean isTraceEnabled() {
        return log.isTraceEnabled();
    }

    @Override
    public void trace(String var1) {
        log.trace(var1);
    }

    @Override
    public void trace(String var1, Throwable var2) {
        log.trace(var1, var2);
    }

    @Override
    public boolean isDebugEnabled() {
        return log.isDebugEnabled();
    }

    @Override
    public void debug(String var1) {
        log.debug(var1);
    }

    @Override
    public void debug(String var1, Throwable var2) {
        log.debug(var1, var2);
    }

    @Override
    public boolean isInfoEnabled() {
        return log.isInfoEnabled();
    }

    @Override
    public void info(String var1) {
        log.info(var1);
    }

    @Override
    public void info(String var1, Throwable var2) {
        log.info(var1, var2);
    }

    @Override
    public boolean isWarnEnabled() {
        return log.isWarnEnabled();
    }

    @Override
    public void warn(String var1) {
        log.warn(var1);
    }

    @Override
    public void warn(String var1, Throwable var2) {
        log.warn(var1, var2);
    }

    @Override
    public boolean isErrorEnabled() {
        return log.isErrorEnabled();
    }

    @Override
    public void error(String var1) {
        log.error(var1);
    }

    @Override
    public void error(String var1, Throwable var2) {
        log.error(var1, var2);
    }
}
